Handle (un)subscribe events

theViz343 opened the pull request:
<!-- See README for documentation, or ask in #zulip-terminal if unclear -->
### What does this PR do, and why?
This PR adds event handling for (un)subscribe events. Without event handling, ZT crashes
if the user narrows onto a message in a stream which is unsubscribed from. This happens mostly 
when a user is logged in from multiple clients, but can also occur if an admin unsubscribes the user.


### Outstanding aspect(s)    <!-- DELETE SECTION IF EMPTY -->
<!-- In what ways is this not fully implemented/functioning? Compared to a discussion/issue? -->
<!-- Do you not understand something? Are you unsure about a certain approach? Want feedback? -->
- [ ] Updated subscribed streams frequently have the wrong/default color. (This seems to be an issue of
ZT not loading the colors.)
- [x] ZT crashes when the currently narrowed message is unsubscribed from.
- [ ] Should we notify user of the new (un)subscription? If yes, then where?(As a popup or as a message in the footer)

### External discussion & connections
<!-- [x] all that apply, specifying topic and adding numbers after # for issues/PRs -->
- [x] Discussed in **#zulip-terminal** in `Adding streams throws key error`
- [x] Fully fixes #816
- [ ] Partially fixes issue #
- [x] Builds upon previous unmerged work in PR #1408 
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?
<!-- [x] all that apply -->
- [x] Manually - Behavioral changes
- [x] Manually - Visual changes
- [x] Adapting existing automated tests
- [x] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [x] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [x] It has a commit summary describing the  motivation and reasoning for the change
- [x] It individually passes linting and tests
- [x] It contains test additions for any new behavior
- [x] It flows clearly from a previous branch commit, and/or prepares for the next commit

### Visual changes    <!-- DELETE SECTION IF NO VISUAL CHANGE -->
<!-- Zulip tips at https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html -->
<!-- For video, try asciinema; after uploading, embed using
-->
[![Preview](https://asciinema.org/a/qSfrl15TUAozL0kmLAb6kmfim.svg)](https://asciinema.org/a/qSfrl15TUAozL0kmLAb6kmfim)

theViz343 committed:
refactor: model: Move date_created field normalization code to function.

This commit moves the date_created field normalization code to a
separate class function called normalize_date_created_field. The
function adds the date_created field for a stream when the server
feature level is less than 30. The value assigned is 'None'.

Test added.

theViz343 committed:
model/api_types: Handle add/remove subscription events.

This commit adds event handling code for newly added/removed
subscriptions. To store the stream subscriptions which are being updated
in the event, a new field named "subscriptions" is added to the
SubscriptionEvent class.

Co-authored-by: Abhirup Pal <abhiruppalmethodist@gmail.com>
Co-authored-by: Vishwesh Pillai <vishwesh103@gmail.com>

theViz343 committed:
model/api_types: Add data structures to support non-subscribed streams.

This commit creates data structures, similar to stream_dict, to support
unsubscribed and never-subscribed streams. These streams are available
in the fetched initial data which is used to populate the new data
structures using the _register_non_subscribed_streams function.

neiljp commented on zulipterminal/model.py:
```diff
@@ -1224,7 +1225,13 @@ def make_reduced_stream_data(stream: Subscription) -> StreamData:
             subscription["color"] = canonicalize_color(subscription["color"])
 
-            self.stream_dict[subscription["stream_id"]] = subscription
+            stream_id = subscription["stream_id"]
+            if stream_id in self.unsubscribed_stream_dict:
+                del self.unsubscribed_stream_dict[stream_id]
+            elif stream_id in self.never_subscribed_stream_dict:
+                del self.never_subscribed_stream_dict[stream_id]
+            self.stream_dict[stream_id] = subscription
```
Minor: since the original line is expanded into a 'synchronize subscriptions' block, it seems clearer to have the stream_id extracted earlier?

That could be achieved by simply separating by a blank line, but the rest of this function could benefit from the use of `stream_id` in any case?

neiljp replied:
I understand that this is aimed at ensuring that the stream 'dicts' are synchronized, and corresponds to the code in the other function, but it seems strange included within the subscription-removal commit, since nothing is calling them except the initial call to this function, when the synchronization *should* be a no-op.

It may be clearer to consider adding and removing subscriptions separately, commit-wise?

neiljp commented on zulipterminal/model.py:
```diff
@@ -1247,6 +1254,29 @@ def make_reduced_stream_data(stream: Subscription) -> StreamData:
+                self.muted_streams.remove(stream_id)
+            if stream_id in self.visual_notified_streams:
+                self.visual_notified_streams.remove(stream_id)
+            if stream_id in self.stream_dict:
+                self.unsubscribed_stream_dict[stream_id] = self.stream_dict[stream_id]
+                del self.stream_dict[stream_id]
```
Readability: for comparison with the subscription function, this should maybe go at the top of the loop?

neiljp commented on zulipterminal/model.py:
```diff
@@ -1446,6 +1446,18 @@ def get_stream_by_id(streams: List[StreamData], stream_id: int) -> StreamData:
+        elif event["op"] == "add":
+            self._subscribe_to_streams(event["subscriptions"])
+            # print(event["subscriptions"])
+            self.normalize_and_cache_message_retention_text()
+            self.normalize_date_created_field()
```
These two lines here suggest these calls may belong in _subscribe_to_streams?

neiljp reviewed (commented):
@theViz343 Thanks for the update on this - there's a lot to process here!

I've left a lot of feedback, but it is often contextual as I was reading through each commit, and in many cases overlaps with other comments to suggest an overall direction to take.

Separate from the specific code suggestions, quite a few of the other comments relate to structure between commits, and adjusting that will make further feedback easier. We can discuss commit order in the stream if that would help.

One additional factor that we may need to handle is unreads, which wasn't handled in the previous PR? I suspect unsubscribing should 'remove' any unreads in a stream, eg. they shouldn't contribute to the all-messages count. However, I don't know whether that data is explicitly cleared on the server side - do we get a mark-read event of some kind? - or whether clients are simply expected to zero any unreads in unsubscribed streams, or simply *hide* any unreads related to was-subscribed streams, ie. where unreads data is available but we are not now subscribed.

theViz343 committed:
model: Add stream id accessor methods.

This commit adds two accessor methods - _get_stream_from_id and
get_all_stream_ids. These two accessor methods help in preparation for
including specific stream property accessor methods in the next set of
commits.

Tests added.

theViz343 committed:
refactor: model: Use get_stream_name instead of stream_dict.

This commit replaces the usage of directly indexing stream_dict for
stream "name" with the new stream property accessor method
"get_stream_name".

Test added.

theViz343 committed:
refactor: model/views: Use get_stream_subscribers instead of stream_dict.

This commit replaces the usage of directly indexing stream_dict for
"subscribers" with the new stream property accessor method
"get_stream_subscribers".

Test added.

theViz343 committed:
refactor: model: Use stream_date_created (get/set)ter methods.

This commit replaces the usage of directly indexing stream_dict for
"date_created" data with the new stream property accessor methods
"get_stream_date_created" and "set_stream_date_created".

Tests added.

theViz343 committed:
refactor: model: Use is_stream_web_public instead of stream_dict.

This commit replaces the usage of directly indexing stream_dict for
"is_web_public" data with the new stream property accessor method
"is_stream_web_public".

Test added.

theViz343 committed:
refactor: model: Use stream_message_retention_days (get/set)ter methods.

This commit replaces the usage of directly indexing stream_dict for
"message_retention_days" data with the new stream property accessor
methods "get_stream_message_retention_days" and
"set_stream_message_retention_days".

Test added.

theViz343 committed:
refactor: model: Use get_subscription_color instead of stream_dict.

This commit replaces the usage of directly indexing stream_dict for
"color" data with the new stream property accessor method
"get_subscription_color".

Test added.

theViz343 committed:
refactor: model: Use is_stream_invite_only instead of stream_dict.

This commit replaces the usage of directly indexing stream_dict for
"is_invite_only" data with the new stream property accessor method
"is_stream_invite_only".

Test added.

theViz343 committed:
model: Add get_all_subscription_ids accessor method.

This commit adds the subscription ids accessor method
"get_all_stream_ids". This helps in preparation for including
subscription color accessor method added in the next commit.

Tests added.

theViz343 committed:
refactor: model/views: Use get_stream_post_policy accessor method.

This commit replaces the usage of directly indexing stream_dict for
"stream_post_policy" with the new stream property accessor method
"get_stream_post_policy".

Test added.

theViz343 committed:
refactor: model/views: Use is_stream_announcement_only accessor method.

This commit replaces the usage of directly indexing stream_dict for
"is_announcement_only" data with the new stream property accessor method
"is_stream_announcement_only".

Test added.

theViz343 committed:
refactor: model/views: Use is_stream_history_public_to_subscribers method.

This commit replaces the usage of directly indexing stream_dict for
"history_public_to_subscribers" data with the new stream property
accessor method "is_stream_history_public_to_subscribers".

Test added.

theViz343 committed:
refactor: model/views: Use get_stream_weekly_traffic method.

This commit replaces the usage of directly indexing stream_dict for
"stream_weekly_traffic" data with the new stream property
accessor method "get_stream_weekly_traffic".

Test added.

theViz343 committed:
refactor: model/views: Use get_stream_rendered_description method.

This commit replaces the usage of directly indexing stream_dict for
"rendered_description" data with the new stream property
accessor method "get_stream_rendered_description".

Test added.

theViz343 committed:
refactor: model/views: Use get_subscription_email instead of stream_dict.

This commit replaces the usage of directly indexing stream_dict for
"email_address" data with the new stream property accessor method
"get_subscription_email".

Test added.

theViz343 committed:
refactor: model: Improve stream_dict typing.

This commit improves the typing of stream_dict by using Subscription
TypedDict instead of Dict[str, Any].

Tests updated.

theViz343 committed:
refactor: model: Rename stream_dict to _subscribed_streams.

This commit renames stream_dict to _subscribed_streams. This change is
necessary due to the introduction of _unsubscribed_streams and
never_subscribed streams in the model.

Tests updated.

theViz343 committed:
model/api_types: Create method in model to handle stream unsubscription.

This commit adds a function named "_unsubscribe_from_streams" to
model.py. This function updates subscription data by removing
unsubscribed streams from class variables - pinned_streams,
unpinned_streams, muted_streams, and visual_notified_streams.

Co-authored-by: Abhirup Pal <abhiruppalmethodist@gmail.com>
Co-authored-by: Vishwesh Pillai <vishwesh103@gmail.com>

theViz343 committed:
model: Remove subscribed stream from nonsubscribed stream data.

The refactor introduced in earlier commits now allows stream data to be
removed from non-subscribed stream data structures, when a stream is
subscribed to.

Test updated.
